libclamav_rust/src/codesign.rs

```diff
@@ -0,0 +1,588 @@
+                                }
+                                Err(Error::InvalidDigitalSignature(m)) => {
+                                    warn!(
+                                        "Invalid digital signature for {:?}: {}",
+                                        signed_file_path, m
+                                    );
+                                    return Err(Error::InvalidDigitalSignature(m));
+                                }
+                                Err(e) => {
+                                    debug!(
+                                        "Error verifying signature with {:?}: {:?}",
+                                        cert_path, e
+                                    );
+
+                                    // Try the next certificate
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            Err(e) => {
+                eprintln!(
+                    "{:?}:{}: Error parsing signature: {}",
+                    signature_file_path, index, e
+                );
+                return Err(Error::CannotVerify(e.to_string()));
+            }
+        };
+    }
+
+    Err(Error::CannotVerify(
+        "Unable to verify any digital signatures".to_string(),
+    ))
+}
+
+pub struct Signer {
+    cert: X509,
+    certs: Stack<X509>,
+    key: PKey<Private>,
+}
+
+impl Signer {
+    pub fn new<P>(
+        key_path: &Path,
+        cert_path: &Path,
+        intermediate_cert_paths: &[P],
+    ) -> Result<Self, Error>
+    where
+        P: AsRef<Path>,
+    {
+        let mut certs: Stack<X509> = Stack::new()?;
+
+        let cert_bytes = std::fs::read(cert_path)?;
+        let cert = X509::from_pem(&cert_bytes)?;
+        debug!("Signing certificate: {:?}", cert);
+        certs.push(cert.clone())?;
+
+        let signing_key_bytes = std::fs::read(key_path)?;
+        let key = PKey::private_key_from_pem(&signing_key_bytes)?;
+        debug!("Signing key: {:?}", key);
+
+        for intermediate_cert_path in intermediate_cert_paths {
+            let intermediate_cert_bytes = std::fs::read(intermediate_cert_path)?;
+            let intermediate_cert = X509::from_pem(&intermediate_cert_bytes)?;
+            debug!("Intermediate certificate: {:?}", &intermediate_cert);
+            certs.push(intermediate_cert)?;
+        }
+
+        Ok(Signer { cert, certs, key })
+    }
+
+    pub fn sign(&self, data: &[u8]) -> Result<Pkcs7, Error> {
+        let flags = Pkcs7Flags::DETACHED | Pkcs7Flags::BINARY;
+
+        Pkcs7::sign(&self.cert, &self.key, &self.certs, data, flags).map_err(Error::OpenSSLError)
+    }
+}
+
+pub struct Verifier {
+    root_ca: X509,
+}
+
+impl Verifier {
+    pub fn new(root_ca_path: &Path) -> Result<Self, Error> {
+        let root_ca_bytes = std::fs::read(root_ca_path)?;
+        let root_ca: X509 = X509::from_pem(&root_ca_bytes)?;
```

Comment:
Should try to read from a pem stack not just a single one, in case it's a bundle. And we should be doing this for the certs_directory for each file in the certs directory. That way the verifier can verify all CVD's.